pkg/socketlb/cgroup.go

```diff
@@ -0,0 +1,205 @@
+		// so we don't need to hold a reference in the agent/bpffs to ensure
+		// the program stays active.
+		log.Debugf("program %s was attached using PROG_ATTACH", name)
+		return nil
+
+	case err != nil:
+		return fmt.Errorf("attach program %s: %w", name, err)
+	}
+
+	// The program was successfully attached using bpf_link, and there was no preexisting pinned link.
+	// The link needs to be pinned and closed.
+	err = l.Pin(filepath.Join(pinPath, name))
+	if err != nil {
+		return fmt.Errorf("pin program %s link at %s: %w", name, pinPath, err)
+	}
+
+	// Successfully created and pinned bpf_link. Closing the link will not remove it, as it is pinned.
+	log.Debugf("program %s attached using bpf_link", name)
+	return l.Close()
+}
+
+// detachCgroup detaches a program with the given name from cgroupRoot. Attempts
+// to open a pinned link with the given name from directory pinPath first,
+// falling back to PROG_DETACH if no pin is present.
+func detachCgroup(name, cgroupRoot, pinPath string) error {
+	pin := filepath.Join(pinPath, name)
+	l, err := link.LoadPinnedLink(pin, &ebpf.LoadPinOptions{})
+	if err == nil {
+		if err := l.Unpin(); err != nil {
+			return fmt.Errorf("unpin link %s: %w", pin, err)
+		}
+		if err := l.Close(); err != nil {
+			return fmt.Errorf("close link %s: %w", name, err)
+		}
+		return nil
+	}
+
+	// No bpf_link pin found, detach all prog_attach progs.
+	log.Debugf("no pinned link '%s', querying cgroup", pin)
+	err = detachAll(attachTypes[name], cgroupRoot)
+	// Treat detaching unsupported attach types as successful.
+	if errors.Is(err, link.ErrNotSupported) {
+		return nil
+	}
+	return err
+}
+
+// detachAll detaches all programs attached to cgroupRoot with the corresponding attach type.
+func detachAll(attach ebpf.AttachType, cgroupRoot string) error {
+	// Query the program ids of all programs currently attached to the given cgroup
+	// with the given attach type. In ciliums case this should always return only one id.
+	ids, err := link.QueryPrograms(link.QueryOptions{
+		Path:   cgroupRoot,
+		Attach: attach,
+	})
+	// We know the cgroup root exists, so EINVAL will likely mean querying
+	// the given attach type is not supported.
+	if errors.Is(err, unix.EINVAL) {
+		err = fmt.Errorf("%s: %w", err, link.ErrNotSupported)
+	}
+	if err != nil {
+		return fmt.Errorf("query cgroup %s for type %s: %w", cgroupRoot, attach, err)
+	}
+
+	if len(ids) == 0 {
+		log.Debugf("no programs in cgroup %s with attach type %s", cgroupRoot, attach)
+		return nil
+	}
+
+	cg, err := os.Open(cgroupRoot)
+	if err != nil {
+		return fmt.Errorf("open cgroup %s: %w", cg.Name(), err)
+	}
+	defer cg.Close()
+
+	// cilium owns the cgroup and assumes only one program is attached.
+	// This allows to remove all ids returned in the query phase.
+	for _, id := range ids {
+		prog, err := ebpf.NewProgramFromID(id)
+		if err != nil {
+			return fmt.Errorf("could not open program id %d: %w", id, err)
```

Comment:
What's the rationale for returning early on a failure? Depending on which program in the list fails, this can lead to non-deterministic state. Any chance of doing a rollback here?

Reply:
Generally there should only be one id in the list, or so cilium assumes. So the rationale is that if we can't retrieve the program from the id, we can't detach, so we fail because otherwise we leave programs behind that should have been removed.